docs/docs/development/new-behavior.md

```diff
@@ -0,0 +1,477 @@
+
+Events can be used similarly to hardware interrupts, through the use of [listeners](#listeners-and-subscriptions).
+
+###### Listeners and Subscriptions
+
+The condensed form of lines 192-225 of the tap-dance driver, shown below, does an excellent job of showcasing the function of listeners and subscriptions with respect to the [ZMK Event Manager](#zmk-event-manager).
+
+```c title="app/src/behaviors/behavior_tap_dance.c (Lines 192-197, 225)"
+static int tap_dance_position_state_changed_listener(const zmk_event_t *eh);
+
+ZMK_LISTENER(behavior_tap_dance, tap_dance_position_state_changed_listener);
+ZMK_SUBSCRIPTION(behavior_tap_dance, zmk_position_state_changed);
+
+static int tap_dance_position_state_changed_listener(const zmk_event_t *eh){
+    // Do stuff...
+}
+```
+
+Listeners, defined by the `ZMK_LISTENER(mod, cb)` function, take in a listener name (`mod`) and a callback function (`cb`) as their parameters. On the other hand subscriptions are defined by the `ZMK_SUBSCRIPTION(mod, ev_type)`, and determine what kind of event (`ev_type`) should invoke the callback function from the listener. In the tap-dance example, this listener executes code depending on a `zmk_position_state_changed` event, or simply, a change in key position. Other types of ZMK events can be found as the name of the `struct` inside each of the files located at `app/include/zmk/events/<Event Type>.h`. All control paths in a listener should `return` one of the [`ZMK_EV_EVENT_*` values](#return-values), which are shown below.
+
+###### `return` values:
+
+- `ZMK_EV_EVENT_BUBBLE`: Keep propagating the event `struct` to the next listener.
+- `ZMK_EV_EVENT_HANDLED`: Stop propagating the event `struct` to the next listener. The event manager still owns the `struct`'s memory, so it will be `free`d automatically. Do **not** free the memory in this function.
+- `ZMK_EV_EVENT_CAPTURED`: Stop propagating the event `struct` to the next listener. The event `struct`'s memory is now owned by your code, so the event manager will not free the event `struct` memory. Make sure your code will release or free the event at some point in the future. (Use the [`ZMK_EVENT_*` macros](#macros) described below.)
+
+###### Macros:
+
+- `ZMK_EVENT_RAISE(ev)`: Start handling this event (`ev`) with the first registered event listener.
+- `ZMK_EVENT_RAISE_AFTER(ev, mod)`: Start handling this event (`ev`) after the event is captured by the named [event listener](#listeners-and-subscriptions) (`mod`). The named event listener will be skipped as well.
+- `ZMK_EVENT_RAISE_AT(ev, mod)`: Start handling this event (`ev`) at the named [event listener](#listeners-and-subscriptions) (`mod`). The named event listener is the first handler to be invoked.
+- `ZMK_EVENT_RELEASE(ev)`: Continue handling this event (`ev`) at the next registered event listener.
+- `ZMK_EVENT_FREE(ev)`: Free the memory associated with the event (`ev`).
+
+#### `DEVICE_DT_INST_DEFINE`
```

Comment:
We should probably include documentation on `DT_INST_FOREACH_STATUS_OKAY` as well along with this section.

Reply:
Sounds good. I just included the `DT_INST_FOREACH_STATUS_OKAY` into the hold-tap code snippet that mentions the `DEVICE_DT_INST_DEFINE` and a simple one-liner explaining that it's a form of error checking. Hope that's sufficient.